src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -204,32 +274,36 @@ static ucs_status_t
 uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
                                     CUdeviceptr *mapped_addr)
 {
+    int key_present             = 0;
     CUmemAccessDesc access_desc = {};
     CUdeviceptr dptr;
     ucs_status_t status;
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
-                CU_MEM_HANDLE_TYPE_FABRIC, 0));
+    status =
+        uct_cuda_ipc_get_rem_mpool_cache(
+                (CUmemFabricHandle*)&key->ph.handle.fabric_handle,
+                &key->ph.pool, &key_present);
     if (status != UCS_OK) {
         return status;
     }
 
     status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+                *key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
     if (status != UCS_OK) {
         return status;
     }
 
-    status = uct_cuda_ipc_init_access_desc(&access_desc);
-    if (status != UCS_OK) {
-        return status;
-    }
+    if (!key_present) {
```

Comment:
If it needs to be called once during hash element creation, then I propose:
- add a function that creates mpool object (uct_cuda_ipc_get_rem_mpool_create), so it calls all these functions inside:
  ucs_malloc
  cuMemPoolImportFromShareableHandle
  uct_cuda_ipc_init_access_desc
  cuMemPoolSetAccess
- call it from `uct_cuda_ipc_get_rem_mpool_cache`
- get rid of `key_present` return arg

Reply:
Another potential flaw is race condition: we lock mutex only during cache creation, but not during setting access. This means that it's possible that from another thread you might retrieve an object from cache, which is constructed but without access set yet.

To overcome this and previous issue I think it makes sense to redesign this function:
```
uct_cuda_ipc_open_memhandle_mempool
    // try to find an element in cache, without updating the cache
    cache = uct_cuda_ipc_find_rem_mpool_cache (lock - unlock)
    if (cache) {
        // This is hot path
        return ImportPointer(cache)
    }
    // cache is not found, construct a new one
    cache = uct_cuda_ipc_create_rem_mpool_cache
                       ucs_malloc
                       cuMemPoolImportFromShareableHandle
                       ImportPointer
                       uct_cuda_ipc_init_access_desc
                       cuMemPoolSetAccess
    // Need to check that cache is not present - may be inserted by another thread
    uct_cuda_ipc_insert_rem_mpool_cache (cache, lock - unlock)
```
Yes, in slow path we lock twice, but for a very short period of time.
In hot path there is just one lock  
This way we:
1. acquire the lock only for the khash get/put operations, and not holding it during the malloc and cuMemPoolImportFromShareableHandle calls
2. Always have fully consistent objects in cache, not half-constructed
3. Update operation is atomic, no stale cache entries possible, no race conditions

Reply:
Ok, now it's consistent and correct, but I'm not sure about performance. Few things to consider:
1. Single lock for the entire time-consuming operation - which means that we block all the other threads even for disjoint requests
2. Spin lock might not be the good choice for long operations, due to busy waiting
3. Global singleton lock (even for readers) might be a source of contention and poor scalability - maybe consider RW lock?

2 important things that I don't know:
- what is the overall delay of the operation in worst case? ~1ms?
- how many different fabric_handle we are supposed to handle? 10/100/1000?

If delay is small and amount of entries is low, then please disregard my comment.
If the overall delay is considerably long, and amount of updates is high, then ideal solution would be: use **RW lock** wiht 2 fine-grained locks:
```
uct_cuda_ipc_open_memhandle_mempool
    // try to find an element in cache, without updating the cache
    // READ lock is scalable across the READERS, after cache is updated we don't
    // even block the readers
    cache = uct_cuda_ipc_find_rem_mpool_cache (READ lock - unlock)
    if (cache) {
        // This is hot path
        return ImportPointer(cache)
    }
    // cache is not found, construct a new one
    cache = uct_cuda_ipc_create_rem_mpool_cache
                       ucs_malloc
                       cuMemPoolImportFromShareableHandle
                       ImportPointer
                       uct_cuda_ipc_init_access_desc
                       cuMemPoolSetAccess
    // Need to check that cache is not present - may be inserted by another thread
    // Only this part is mutually exclusive = blocks all threads, but for a very short period of time
    uct_cuda_ipc_insert_rem_mpool_cache (cache, WRITE lock - unlock)
```

Reply:
@iyastreb Do the latest changes look good?

Reply:
@iyastreb 
>> 2 important things that I don't know:
    what is the overall delay of the operation in worst case? ~1ms?
    how many different fabric_handle we are supposed to handle? 10/100/1000?

Worst case will be ~1ms but it depends on driver internal operations. This cache is expected to handle in the range of 10-100 fabric handles with majority operations being reads rather than updates. 

Thanks for RW lock usage suggestion. I'm not sure if khash datastructure is thread-safe where one of the threads could be updating the hashmap while others can read unrelated keys and associated values (for eg, what happens if one of the updates needs to allocate more memory and update hashmap metadata?). For this reason, in commit https://github.com/openucx/ucx/pull/10082/commits/fa00e3129627e2efdc5d1abbda2581c931bbd6be, I switched to using a single RW lock for readers and writers. All readers use rdlock to check if key is in hash and use imported pool to generate mapped pointer if key is present. If not, one writer uses wrlock to update the hash and use imported mpool for mapped pointer. This doesn't allow concurrency between readers and writers of unrelated keys but it should permit multiple readers to benefit from not waiting for other reads to complete while there are no writes. 

Let me know if the changes look good and if you have other thoughts.

Reply:
Thank you for addressing my comments! Now it is safe.
You are right, khash impl is not thread-safe for disjoint updates, so in our case a writer blocks all the readers, which is expected. For disjoint access parallelism we would need to use concurrent data structures (or STM) which we don't have at disposal